src/main/kotlin/com/chloe/kotlinserv/MainVerticle.kt

```diff
@@ -1,23 +1,97 @@
+    conf.jdbcUrl = url
+    conf.driverClassName = "com.clickhouse.jdbc.ClickHouseDriver"
+    conf.username = config.getString("clickhouse.user")
+    conf.password = config.getString("clickhouse.password")
+
+    val ds = HikariDataSource(conf)
+
+    val port = config.getInt("port")
     val myServer = VertxHttpServer()
     val json = Gson()
 
     val getRoute = HttpRoute("/countrystats", HttpMethod.GET) {
-        val data = CountryStats("23-12-2000", "RUS", 3)
-        HttpResponse(200, responseBody=json.toJson(data), mapOf("content-type" to "application/json"))
+        val connection = ds.connection
+
+        connection.use {
+            val statement =
+                connection.prepareStatement("SELECT country, toDate(timestamp) as time, count(country) FROM chloe.events WHERE country='user' group by country, toDate(timestamp)")
+            val result = statement.executeQuery()
+
+            val list = mutableListOf<CountryStats>()
+            while (result.next()) {
+                list.add(
+                    CountryStats(
+                        data = result.getString(1),
+                        country = result.getString(2),
+                        count = result.getInt(3)
+                    )
+                )
+            }
+
+            if (list.isEmpty()) {
+                HttpResponse(
+                    code = 204,
+                    responseBody = "No content",
+                    contentType = mapOf("content-type" to "text/plain")
+                )
+            } else {
+                HttpResponse(
+                    code = 200,
+                    responseBody = json.toJson(list),
+                    contentType = mapOf("content-type" to "application/json")
+                )
+            }
+        }
     }
 
+    val query = "insert into chloe.events (timestamp, country, ipAddress, userId) values (?, ?, ?, ?)"
+    val batch = Batch(ds = ds, query = query)
+
     val postRoute = HttpRoute("/geodata", HttpMethod.POST) {
-        val data = GeoData("RUS", 3405454052, "user1")
-        val jsonContent = json.toJson(data)
-        HttpResponse(200, jsonContent, mapOf("content-type" to "application/json"))
+        try {
+            val data = json.fromJson(it.body, GeoData::class.java)
+            val requestHeaders = it.requestHeaders["x-forwarded-for"]
+
+            batch.addToList(data, requestHeaders?.firstOrNull())
+
+            HttpResponse(
+                code = 200,
+                responseBody = null,
+                contentType = mapOf("content-type" to "application/json")
+            )
+        } catch (e: java.lang.NullPointerException) {
+            HttpResponse(
+                code = 400,
+                responseBody = "Bad request",
+                contentType = mapOf("content-type" to "text/plain")
+            )
+        }
     }
+    myServer.start(port, listOf(getRoute, postRoute))
+
+    val executor = Executors.newScheduledThreadPool(1)
```

Comment:
lets move executor and scheduleAtFixedRate process to your Batch class

should be something like it in your class

init {
executor.scheduleAtFixedRate(
        { run() },
        20,
        20,
        TimeUnit.SECONDS
    )
}